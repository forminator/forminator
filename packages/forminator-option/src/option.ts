import { FirmMap } from './firm-map';

export const FORMINATOR_OPTION = Symbol('FORMINATOR_OPTION');
export type Defined = {} | null;
export type Some<Value extends Defined> = {
  some: true;
  value: Value;
} & OptionFns<Value>;
export type None<Value extends Defined> = {
  some: false;
} & OptionFns<Value>;
export type Option<Value extends Defined> = Some<Value> | None<Value>;

export interface OptionFns<Value extends Defined> {
  readonly [FORMINATOR_OPTION]: true;

  unwrap(this: Option<Value>): Value;

  unwrap_or(this: Option<Value>, value: Value): Value;

  or(this: Option<Value>, value: Option<Value>): Option<Value>;

  map<T>(this: Option<Value>, fn: (value: Value) => T): Option<T & Defined>;

  isSome(this: Option<Value>): this is Some<Value>;

  isNone(this: Option<Value>): this is None<Value>;
}

function getOptionFns<Value extends Defined>(): OptionFns<Value> {
  return {
    [FORMINATOR_OPTION]: true,
    unwrap() {
      return unwrap(this);
    },
    unwrap_or(value: Value): Value {
      return unwrap_or(this, value);
    },
    or(this: Option<Value>, value: Option<Value>) {
      return or(this, value);
    },
    map<T>(this: Option<Value>, fn: (value: Value) => T) {
      return mapOption(this, fn);
    },
    isSome() {
      return isSome(this);
    },
    isNone(this: Option<Value>) {
      return isNone(this);
    },
  };
}

const optionFns = getOptionFns<any>();

function createSome<Value extends Defined>(
  value: Value,
): Some<Value> & OptionFns<Value> {
  if (value === undefined) {
    throw new Error("some value can't be undefined");
  }
  return Object.assign(Object.create(optionFns), { some: true, value });
}

function createNone<Value extends Defined>(): None<Value> {
  return Object.assign(Object.create(optionFns), { some: false });
}

const theNone = createNone<any>();

const cache = new FirmMap<any, Some<any>>();

export function some<Value extends Defined>(
  value: Value,
): Some<Value> & OptionFns<Value> {
  if (!cache.has(value)) {
    cache.set(value, createSome(value));
  }
  return cache.get(value)!;
}

export function none<Value extends Defined>(): None<Value> {
  return theNone;
}

export function isOption(value: any): value is Option<unknown & Defined> {
  return !!(value && value[FORMINATOR_OPTION]);
}

function isSome<Value extends Defined>(
  option: Option<Value>,
): option is Some<Value> & OptionFns<Value> {
  return option.some;
}

function isNone<Value extends Defined>(
  option: Option<Value>,
): option is None<Value> {
  return !option.some;
}

export class NoneError extends Error {
  constructor() {
    super(`value is None.`) /* istanbul ignore next */;

    this.name = 'NoneError';
    Object.setPrototypeOf(this, NoneError.prototype);
  }
}

function unwrap<Value extends Defined>(option: Option<Value>): Value {
  if (isSome(option)) {
    return option.value;
  }
  throwNoneError();
}

function unwrap_or<Value extends Defined>(
  option: Option<Value>,
  value: Value,
): Value {
  if (isSome(option)) {
    return option.value;
  }
  return value;
}

function or<Value extends Defined>(
  option: Option<Value>,
  value: Option<Value>,
): Option<Value> {
  if (isSome(option)) {
    return option;
  }
  return value;
}

export function isNoneError(e: any): e is NoneError {
  return e instanceof NoneError;
}

export function throwNoneError(): never {
  throw new NoneError();
}

export function catchNoneError<R>(fn: () => R): Option<R & Defined> {
  try {
    return intoOption(fn());
  } catch (e) {
    if (isNoneError(e)) {
      return none();
    }
    throw e;
  }
}

function mapOption<V extends Defined, T>(
  option: Option<V>,
  fn: (v: V) => T,
): Option<T & Defined> {
  if (isNone(option)) {
    return none();
  }
  return intoOption(fn(option.value));
}

export function intoOption<Value>(
  value: Value | undefined,
): Option<Value & Defined> {
  if (value === undefined) {
    return none();
  }
  return some(value);
}

export function fromOption<Value extends Defined>(
  option: Option<Value>,
): Value | undefined {
  return option.isSome() ? option.value : undefined;
}
