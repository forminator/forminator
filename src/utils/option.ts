export type Some<Value> = { some: true; value: Value } & OptionFns<Value>;
export type None<Value> = { some: false } & OptionFns<Value>;
export type Option<Value> = Some<Value> | None<Value>;

export interface OptionFns<Value> {
  ok(this: Option<Value>): Value;
  or(this: Option<Value>, value: Value): Value;
  map<T>(this: Option<Value>, fn: (value: Value) => T): Option<T>;
  isSome(this: Option<Value>): this is Some<Value>;
  isNone(this: Option<Value>): this is None<Value>;
}

function getOptionFns<Value>(): OptionFns<Value> {
  return {
    ok() {
      return ok(this);
    },
    or(this: Option<Value>, value: Value) {
      return or(this, value);
    },
    map<T>(this: Option<Value>, fn: (value: Value) => T) {
      return mapOption(this, fn);
    },
    isSome() {
      return isSome(this);
    },
    isNone(this: Option<Value>) {
      return isNone(this);
    },
  };
}

const optionFns = getOptionFns<any>();

export function some<Value>(value: Value): Some<Value> & OptionFns<Value> {
  return Object.assign(Object.create(optionFns), { some: true, value });
}
export function none<Value>(): None<Value> {
  return Object.assign(Object.create(optionFns), { some: false });
}
function isSome<Value>(
  option: Option<Value>,
): option is Some<Value> & OptionFns<Value> {
  return option.some;
}
function isNone<Value>(option: Option<Value>): option is None<Value> {
  return !option.some;
}

export class NoneError extends Error {
  constructor() {
    super(`value is None.`) /* istanbul ignore next */;

    this.name = 'NoneError';
    Object.setPrototypeOf(this, NoneError.prototype);
  }
}

function ok<Value>(option: Option<Value>): Value {
  if (isSome(option)) {
    return option.value;
  }
  throw new NoneError();
}

function or<Value>(option: Option<Value>, value: Value): Value {
  if (isSome(option)) {
    return option.value;
  }
  return value;
}

function isNoneError(e: any): e is NoneError {
  return e instanceof NoneError;
}

export function catchNoneError<R>(fn: () => R): Option<R> {
  try {
    return some(fn());
  } catch (e) {
    if (isNoneError(e)) {
      return none();
    }
    throw e;
  }
}

function mapOption<V, T>(option: Option<V>, fn: (v: V) => T): Option<T> {
  if (isNone(option)) {
    return none();
  }
  return some(fn(option.value));
}
