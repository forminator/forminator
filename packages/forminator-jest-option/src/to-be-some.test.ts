import { none, some } from '@forminator/option';

describe('.toBeSome', () => {
  it('should pass on some', () => {
    expect(some(5)).toBeSome();
  });
  it('should pass on some with expected value', function () {
    expect(some(5)).toBeSome(5);
    const value = { foo: 5 };
    expect(some(value)).toBeSome(value);
  });
  it('should fail on none', () => {
    expect(() => expect(none()).toBeSome()).toThrowErrorMatchingSnapshot();
  });
  it('should fail on none with expected value', () => {
    expect(() => expect(none()).toBeSome(5)).toThrowErrorMatchingSnapshot();
    expect(() =>
      expect(none()).toBeSome({ foo: 5 }),
    ).toThrowErrorMatchingSnapshot();
  });
  it('should fail on some with different value', () => {
    expect(() => expect(some(6)).toBeSome(5)).toThrowErrorMatchingSnapshot();
    expect(() =>
      expect(some({ foo: 5 })).toBeSome({ foo: 6 }),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      expect(some({ foo: 5 })).toBeSome({ foo: 5 }),
    ).toThrowErrorMatchingSnapshot();
  });

  it('should fail on other value', function () {
    expect(() => expect(5).toBeSome()).toThrowErrorMatchingSnapshot();
  });
  it('should fail on other value with expected value', function () {
    expect(() => expect(5).toBeSome(5)).toThrowErrorMatchingSnapshot();
    const value = { foo: 5 };
    expect(() => expect(value).toBeSome(value)).toThrowErrorMatchingSnapshot();
  });
});

describe('.not.toBeSome', () => {
  it('should fail on some', () => {
    expect(() => expect(some(5)).not.toBeSome()).toThrowErrorMatchingSnapshot();
  });
  it('should fail on some with expected value', function () {
    expect(() =>
      expect(some(5)).not.toBeSome(5),
    ).toThrowErrorMatchingSnapshot();
    const value = { foo: 5 };
    expect(() =>
      expect(some(value)).not.toBeSome(value),
    ).toThrowErrorMatchingSnapshot();
  });
  it('should pass on none', () => {
    expect(none()).not.toBeSome();
  });
  it('should pass on none with expected value', () => {
    expect(none()).not.toBeSome(5);
    expect(none()).not.toBeSome({ foo: 5 });
  });
  it('should pass on some with different value', () => {
    expect(some(6)).not.toBeSome(5);
    expect(some({ foo: 5 })).not.toBeSome({ foo: 6 });
    expect(some({ foo: 5 })).not.toBeSome({ foo: 5 });
  });

  it('should pass on other value', function () {
    expect(5).not.toBeSome();
  });
  it('should pass on other value with expected value', function () {
    expect(5).not.toBeSome(5);
    const value = { foo: 5 };
    expect(value).not.toBeSome(value);
  });
});
