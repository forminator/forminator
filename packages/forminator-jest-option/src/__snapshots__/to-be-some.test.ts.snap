// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toBeSome should fail on some 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.toBeSome()[22m

Expected: not [32msome([32mAnything[39m[32m)[39m
Received: [31msome([31m5[39m[31m)[39m
"
`;

exports[`.not.toBeSome should fail on some with expected value 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.toBeSome([22m[32mexpected[39m[2m)[22m

Expected: not [32msome([32m5[39m[32m)[39m
"
`;

exports[`.not.toBeSome should fail on some with expected value 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.toBeSome([22m[32mexpected[39m[2m)[22m

Expected: not [32msome([32m{\\"foo\\": 5}[39m[32m)[39m
"
`;

exports[`.toBeSome should fail on none 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome()[22m

Expected: [32msome([32mAnything[39m[32m)[39m
Received: [31mnone()[39m
"
`;

exports[`.toBeSome should fail on none with expected value 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m)[22m

Expected: [32msome([32m5[39m[32m)[39m
Received: [31mnone()[39m
"
`;

exports[`.toBeSome should fail on none with expected value 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m)[22m

Expected: [32msome([32m{\\"foo\\": 5}[39m[32m)[39m
Received: [31mnone()[39m
"
`;

exports[`.toBeSome should fail on other value 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome()[22m

Expected: [32msome([32mAnything[39m[32m)[39m
Received: [31m5[39m
"
`;

exports[`.toBeSome should fail on other value with expected value 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m)[22m

Expected: [32msome([32m5[39m[32m)[39m
Received: [31m5[39m
"
`;

exports[`.toBeSome should fail on other value with expected value 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m)[22m

Expected: [32msome([32m{\\"foo\\": 5}[39m[32m)[39m
Received: [31m{\\"foo\\": 5}[39m
"
`;

exports[`.toBeSome should fail on some with different value 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m) // some with different value[22m

Expected: [32m5[39m
Received: [31m6[39m
"
`;

exports[`.toBeSome should fail on some with different value 2`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m) // some with different value[22m

[32m- Expected  - 1[39m
[31m+ Received  + 1[39m

[2m  Object {[22m
[32m-   \\"foo\\": 6,[39m
[31m+   \\"foo\\": 5,[39m
[2m  }[22m
"
`;

exports[`.toBeSome should fail on some with different value 3`] = `
"[2mexpect([22m[31mreceived[39m[2m).toBeSome([22m[32mexpected[39m[2m) // some with different value[22m

Expected: [32m{\\"foo\\": 5}[39m
Received: serializes to the same string
"
`;
